cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringAgentService.java

```diff
@@ -412,6 +440,59 @@ private PeerProgramUpdates getPeerProgramUpdates(String afterMessageId) {
     return new PeerProgramUpdates(peerToNotifications, lastMessageId);
   }
 
+  /**
+   * Reads from per-peer topic and returns program status updates for the given tethered peer.
+   */
+  private ImmutablePair<List<Notification>, String> getProgramStateUpdates(String peerName) {
+    String topic = programStateTopicPrefix + peerName;
+    String lastMessageId = lastMessageIdForPeer.get(peerName);
+    List<Notification> programUpdates = new ArrayList<>();
+    try (CloseableIterator<Message> iterator = messageFetcher.fetch(NamespaceId.SYSTEM.getNamespace(),
+                                                                    topic,
+                                                                    programUpdateFetchSize,
+                                                                    lastMessageIdForPeer.get(peerName))) {
+      while (iterator.hasNext()) {
+        Message message = iterator.next();
+        programUpdates.add(message.decodePayload(r -> GSON.fromJson(r, Notification.class)));
+        lastMessageId = message.getId();
+      }
+    } catch (TopicNotFoundException e) {
+      createTopicIfNeeded(new TopicId(NamespaceId.SYSTEM.getNamespace(), topic));
+    } catch (IOException e) {
+      LOG.error("Exception when fetching program updates from per-peer topic. Will retry again during next poll", e);
+    }
+    return ImmutablePair.of(programUpdates, lastMessageId);
+  }
+
+  private void persistNotifications(PeerProgramUpdates peerProgramUpdates, List<PeerInfo> peers) throws IOException {
+    for (PeerInfo peer: peers) {
+      String peerName = peer.getName();
+      List<String> notifications = peerProgramUpdates.peerToNotifications.getOrDefault(peerName, new ArrayList<>())
+        .stream().map(n -> GSON.toJson(n)).collect(Collectors.toList());
+      if (notifications.isEmpty()) {
+        continue;
+      }
+      String topic = programStateTopicPrefix + peerName;
+      try {
+        messagePublisher.publish(NamespaceId.SYSTEM.getNamespace(), topic,
+                                 notifications.toArray(new String[0]));
+      } catch (TopicNotFoundException e) {
+        createTopicIfNeeded(new TopicId(NamespaceId.SYSTEM.getNamespace(), topic));
+      }
+    }
+  }
+
+  private void createTopicIfNeeded(TopicId topicId) {
+    LOG.debug("Creating topic {}", topicId.getTopic());
```

Comment:
should we say attempting to create OR move this after `createTopic` so that we can say x topic created

Reply:
changed to Created topic x after creating topic